src/execute.rs

```diff
@@ -0,0 +1,654 @@
+    let delta_y = calc_amount_1_one_unit_liq(lower_tick, current_tick)?;
+    let total_delta = delta_x.checked_add(delta_y)?;
+    let asset0_ratio = delta_x.checked_div(total_delta)?;
+    let asset1_ratio = Decimal256::one().checked_sub(asset0_ratio)?;
+
+    Ok((asset0_ratio, asset1_ratio))
+}
+
+fn get_next_sqrt_price_from_amount0_in_round_up(
+    liquidity: Decimal256,
+    sqrt_price_current: Decimal256,
+    token_in: Decimal256,
+) -> Decimal256 {
+    let numerator = liquidity * sqrt_price_current;
+    let denominator = liquidity + (token_in * sqrt_price_current);
+    numerator / denominator
+}
+
+fn get_next_sqrt_price_from_amount1_in_round_down(
+    liquidity: Decimal256,
+    sqrt_price_current: Decimal256,
+    token_in: Decimal256,
+) -> Decimal256 {
+    sqrt_price_current + (token_in / liquidity)
+}
+
+fn calc_amount_one_delta(
+    liquidity: Decimal256,
+    mut sqrt_price_current: Decimal256,
+    mut sqrt_price_next: Decimal256,
+    should_round_up: bool,
+) -> Decimal256 {
+    if sqrt_price_next > sqrt_price_current {
+        std::mem::swap(&mut sqrt_price_current, &mut sqrt_price_next);
+    }
+    let result = liquidity * (sqrt_price_current.abs_diff(sqrt_price_next));
+    match should_round_up {
+        true => result.ceil(),
+        false => result,
+    }
+}
+
+fn calc_amount_zero_delta(
+    liquidity: Decimal256,
+    mut sqrt_price_current: Decimal256,
+    mut sqrt_price_next: Decimal256,
+    should_round_up: bool,
+) -> Decimal256 {
+    if sqrt_price_next > sqrt_price_current {
+        std::mem::swap(&mut sqrt_price_current, &mut sqrt_price_next);
+    }
+    let mul1 = liquidity * (sqrt_price_current - sqrt_price_next);
+    let result = mul1 / (sqrt_price_current * sqrt_price_next);
+    match should_round_up {
+        true => result.ceil(),
+        false => result,
+    }
+}
+
+fn iterative_naive_approach(
+    deps: &mut DepsMut,
+    // liquidity: Decimal,
+    // token_out_denom: String,
+    // sqrt_price_current: Decimal,
+    // sqrt_price_next: Decimal,
+    token_to_swap: Coin,
+    token_initially_provided: Coin,
+    pool: Pool,
+    upper_tick: i64,
+    lower_tick: i64,
+) -> Result<Coin, ContractError> {
+    // We are assuming here we have enough liquidity to swap the entire amount of token_in
+    // Will change later
+    let token_0_in = token_to_swap.denom == pool.token0;
+    let token_1_in = token_to_swap.denom == pool.token1;
+    let token_provided_dec =
+        Decimal256::from_str(token_initially_provided.amount.to_string().as_str())?;
+
+    let original = pool.current_sqrt_price.to_string();
+    let parts: Vec<&str> = original.split('.').collect();
+    let integer_part = parts[0];
+    let fractional_part = &parts[1][0..18]; // Take only first 18 digits after the decimal point
+    let rounded = format!("{}.{}", integer_part, fractional_part);
+
+    let pool_cur_sqrt_price = Decimal256::from_str(&rounded).unwrap();
```

Comment:
maybe we check to make sure it is not zero and fail if so to handle my comment above